cpp/src/arrow/util/memory-pool.cc

```diff
@@ -74,15 +74,22 @@ Status InternalMemoryPool::Allocate(int64_t size, uint8_t** out) {
   return Status::OK();
 }
 
-int64_t InternalMemoryPool::bytes_allocated() const {
+uint64_t InternalMemoryPool::bytes_allocated() const {
   std::lock_guard<std::mutex> guard(pool_lock_);
   return bytes_allocated_;
 }
 
-void InternalMemoryPool::Free(uint8_t* buffer, int64_t size) {
+Status InternalMemoryPool::Free(uint8_t* buffer, int64_t size) {
   std::lock_guard<std::mutex> guard(pool_lock_);
+  if (bytes_allocated_ < size) {
+    std::stringstream ss;
+    ss << "free of size " << size << " larger than allocated bytes " << bytes_allocated_ << " failed ";
+    return Status::Invalid(ss.str());
+  }
```

Comment:
Thanks guys for your comment. I changed to DCHECK. 
In addition, I agree on that we need a better interface. I'll further look at this.